### ISSUE TITLE
Issue with UnitSetXY and City:GetGarrisonedUnit()

### ISSUE THREAD

**Issue by Chrisy15** [LoneGazebo/Community-Patch-DLL#2892]
_1. Mod version (i.e Date - 4/23):_

12/5 

_2. Mod list (if using Vox Populi only, leave blank):_
Core CP
IGE
Cyrus v3 (shouldn't affect issue)

_3. Error description:_ 
I'm trying to use the method GetGarrisonedUnit on the GameEvent UnitSetXY. In vanilla, everything works fine; however, when using the CP GetGarrisonedUnit seems to be out-of-sync with the unit's actual position.

Function:
```
function C15_Darius_GarrisonHappiness(playerID, unitID, iX, iY)
	local pPlayer = Players[playerID]
	if HasTrait(pPlayer, traitIDDarius) then
		print("Trait firing")
		print("iX = ", iX, "iY = ", iY)
		for pCity in pPlayer:Cities() do
			for i = 0, pCity:Plot():GetNumUnits() - 1 do
				local pUnit = pCity:Plot():GetUnit(i)
				print(pUnit:GetUnitType())
			end
			print("pCity:GetX = ", pCity:GetX(), "pCity:GetY = ", pCity:GetY())
			if Map.GetPlot(iX, iY) == pCity:Plot() then 
				print("The plot that XY points to is the city's plot") 
			else 
				print ("The plot XY points to is not the city's plot") 
			end
			print("Iterating cities")
			print("GetGarrisonedUnit:", pCity:GetGarrisonedUnit())
			local pUnitIDUnit = pPlayer:GetUnitByID(unitID)
			print("pUnitIDUnit:GetX = ", pUnitIDUnit:GetX(), "pUnitIDUnit:GetY = ", pUnitIDUnit:GetY())
			if pCity:GetGarrisonedUnit() then
				print("pGarrison:GetX = ", pCity:GetGarrisonedUnit():GetX(), "pGarrison:GetY", pCity:GetGarrisonedUnit():GetY())
				pCity:SetNumRealBuilding(iTraitHappinessDummy, pCity:GetGarrisonedUnit():GetLevel())
				print("NumBuildig:", pCity:GetNumRealBuilding(iTraitHappinessDummy))
			else
				pCity:SetNumRealBuilding(iTraitHappinessDummy, 0)
				print("NumBuilding:", pCity:GetNumRealBuilding(iTraitHappinessDummy))
			end
		end
	end
end

if isPersiaCivActive then
	GameEvents.UnitSetXY.Add(C15_Darius_GarrisonHappiness)
end
```

Vanilla:
 
```
City settle:
C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	-2147483647	iY = 	-2147483647
 C15_Darius_Lua: pCity:GetX = 	32	pCity:GetY = 	39
 C15_Darius_Lua: The plot XY points to is not the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	nil
 C15_Darius_Lua: pUnitIDUnit:GetX = 	-2147483647	pUnitIDUnit:GetY = 	-2147483647
 C15_Darius_Lua: NumBuilding:	0

Move starting Warrior into Capital:
 C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	32	iY = 	39
 C15_Darius_Lua: 83
 C15_Darius_Lua: pCity:GetX = 	32	pCity:GetY = 	39
 C15_Darius_Lua: The plot that XY points to is the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	table: 615DDF28
 C15_Darius_Lua: pUnitIDUnit:GetX = 	32	pUnitIDUnit:GetY = 	39
 C15_Darius_Lua: pGarrison:GetX = 	32	pGarrison:GetY	39
 C15_Darius_Lua: NumBuildig:	1

Move starting Warrior out of capital:
 C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	33	iY = 	40
 C15_Darius_Lua: pCity:GetX = 	32	pCity:GetY = 	39
 C15_Darius_Lua: The plot XY points to is not the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	nil
 C15_Darius_Lua: pUnitIDUnit:GetX = 	33	pUnitIDUnit:GetY = 	40
 C15_Darius_Lua: NumBuilding:	0 
```

CP:

```
City Settle:
 C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	-2147483647	iY = 	-2147483647
 C15_Darius_Lua: pCity:GetX = 	52	pCity:GetY = 	28
 C15_Darius_Lua: The plot XY points to is not the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	nil
 C15_Darius_Lua: pUnitIDUnit:GetX = 	-2147483647	pUnitIDUnit:GetY = 	-2147483647
 C15_Darius_Lua: NumBuilding:	0

Move starting Warrior into Capital:
 C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	52	iY = 	28
 C15_Darius_Lua: 83
 C15_Darius_Lua: pCity:GetX = 	52	pCity:GetY = 	28
 C15_Darius_Lua: The plot that XY points to is the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	nil
 C15_Darius_Lua: pUnitIDUnit:GetX = 	52	pUnitIDUnit:GetY = 	28
 C15_Darius_Lua: NumBuilding:	0

Move starting Warrior out of Capital:
 C15_Darius_Lua: Trait firing
 C15_Darius_Lua: iX = 	51	iY = 	29
 C15_Darius_Lua: pCity:GetX = 	52	pCity:GetY = 	28
 C15_Darius_Lua: The plot XY points to is not the city's plot
 C15_Darius_Lua: Iterating cities
 C15_Darius_Lua: GetGarrisonedUnit:	table: 6CFE7AC0
 C15_Darius_Lua: pUnitIDUnit:GetX = 	51	pUnitIDUnit:GetY = 	29
 C15_Darius_Lua: pGarrison:GetX = 	51	pGarrison:GetY	29
 C15_Darius_Lua: NumBuildig:	1 
```

So as you can see, with Vanilla the GetGarrisonedUnit method returns the correct value - the ID when the unit moves into the city, and nil when it moves out.

The CP on the other hand seems to not update the method - despite UnitSetXY saying that the unit's in the city, Plot:GetNumUnits() saying there's a unit on the City tile, and the _unit_ saying that it's in the city, GetGarrisonedUnit disagrees; when everything says the unit has left the city, the method believes it is still in there.

_4. Steps to reproduce (optional):_
Multiple hours of testing with LeeS

**Comment by LoneGazebo** [comment]
@ilteroi any ideas?

**Comment by ilteroi** [comment]
can't claim to understand those logs, please make it a bit more readable?

anyway i put some logs of my own and it's working fine:

phalanx passing through
garrison in Athens is now 0x66629AF0
garrison in Athens is now 0x00000000

scout passing through
garrison in Athens is now 0x666231B8
garrison in Athens is now 0x00000000

archer passing through
garrison in Athens is now 0x6CAA41B0
garrison in Athens is now 0x00000000

**Comment by Chrisy15** [comment]
What is it that's unclear about them? Too few print statements? Too many?

How are you getting those results? I've always found GetGarrisonedUnit to return nil if there's no unit, or a pointer if there is; how have you ended up with a hex number?

**Comment by ilteroi** [comment]
that's an additional log message i put into CvCity::SetGarrison(CvUnit* pUnit)

```
	OutputDebugString( CvString::format( "garrison in %s is now 0x%08X\n", getName().c_str(), pUnit ).c_str() );
```

**Comment by Chrisy15** [comment]
Too late to be discussing this tbh (almost one in the morning - perfect time for civ modding! :p), but I think I might've found something.
Inside CvUnit.cpp, I managed to find the define for UnitSetXY - and by the looks of things, the GameEvent fires before the garrison logic is done.
See screenshot for (slightly) more details/explanation.
![image](https://cloud.githubusercontent.com/assets/24508935/21085043/85363364-c004-11e6-923f-341a445d708d.png)
The way I see it, the Lua hook will fire (thus changing the unit's X/Y), but the garrison status won't yet have been changed. Thus, despite the unit being on the city's tile, the city will return as having no garrison. Then after the hook has been called, the garrison stuff is calculated (which works - because from then on the city will show as having the unit as its garrison). This would also work for leaving the city, because that is also handled after the Lua hook is called.

Am I barking up the wrong tree?
